code/modules/modular_computers/file_system/programs/techweb.dm

```diff
@@ -0,0 +1,218 @@
+		if (node.discount_experiments?.len)
+			node_cache[cid]["discount_experiments"] = node.discount_experiments
+
+	// Build design cache
+	var/design_cache = list()
+	var/datum/asset/spritesheet/research_designs/spritesheet = get_asset_datum(/datum/asset/spritesheet/research_designs)
+	var/size32x32 = "[spritesheet.name]32x32"
+	for (var/did in SSresearch.techweb_designs)
+		var/datum/design/design = SSresearch.techweb_designs[did] || SSresearch.error_design
+		var/cid = "[compress_id(design.id)]"
+		var/size = spritesheet.icon_size_id(design.id)
+		design_cache[cid] = list(
+			design.name,
+			"[size == size32x32 ? "" : "[size] "][design.id]"
+		)
+
+	// Ensure id cache is included for decompression
+	var/flat_id_cache = list()
+	for (var/id in id_cache)
+		flat_id_cache += id
+
+	.["static_data"] = list(
+		"node_cache" = node_cache,
+		"design_cache" = design_cache,
+		"id_cache" = flat_id_cache
+	)
+
+/**
+ * Compresses an ID to an integer representation using the id_cache, used for deduplication
+ * in sent JSON payloads
+ *
+ * Arguments:
+ * * id - the ID to compress
+ */
+/datum/computer_file/program/science/proc/compress_id(id)
+	if (!id_cache[id])
+		id_cache[id] = id_cache_seq++
+	return id_cache[id]
+
+/datum/computer_file/program/science/proc/research_node(id, mob/user)
+	if(!stored_research.available_nodes[id] || stored_research.researched_nodes[id])
+		computer.say("Node unlock failed: Either already researched or not available!")
+		return FALSE
+	var/datum/techweb_node/tech_node = SSresearch.techweb_node_by_id(id)
+	if(!istype(tech_node))
+		computer.say("Node unlock failed: Unknown error.")
+		return FALSE
+	var/list/price = tech_node.get_price(stored_research)
+	if(stored_research.can_afford(price))
+		computer.investigate_log("[key_name(user)] researched [id]([json_encode(price)]) on techweb id [stored_research.id] via [computer].", INVESTIGATE_RESEARCH)
+		if(stored_research == SSresearch.science_tech)
+			SSblackbox.record_feedback("associative", "science_techweb_unlock", 1, list("id" = "[id]", "name" = tech_node.display_name, "price" = "[json_encode(price)]", "time" = SQLtime()))
+		if(stored_research.research_node_id(id))
+			computer.say("Successfully researched [tech_node.display_name].")
+			var/logname = "Unknown"
+			if(isAI(user))
+				logname = "AI: [user.name]"
+			if(iscarbon(user))
+				var/obj/item/card/id/idcard = user.get_active_held_item()
+				if(istype(idcard))
+					logname = "User: [idcard.registered_name]"
+			if(ishuman(user))
+				var/mob/living/carbon/human/H = user
+				var/obj/item/worn = H.wear_id
+				if(istype(worn))
+					var/obj/item/card/id/ID = worn.GetID()
+					if(istype(ID))
+						logname = "User: [ID.registered_name]"
+			var/sci_log_len = stored_research.research_logs.len
+			stored_research.research_logs += null
+			stored_research.research_logs[++sci_log_len] = list(tech_node.display_name, price["General Research"], logname, "[get_area(src)] ([computer.x],[computer.y],[computer.z])")
```

Comment:
```suggestion
			stored_research.research_logs += list(list(tech_node.display_name, price["General Research"], logname, "[get_area(src)] ([computer.x],[computer.y],[computer.z])"))
```

I know MSO does this but I honestly hate it, it should just be a macro if it's that important.

Reply:
And at this point, remove `sci_log_len` as well, and probably that `+= null` too..